django_components/component_registry.py

```diff
@@ -11,11 +14,16 @@ def __init__(self):
         self._registry = {}  # component name -> component_class mapping
 
     def register(self, name=None, component=None):
-        if name in self._registry:
+        if (
+            name in self._registry
+            and name
+            and component
+            and inspect.getsource(component)
+            != inspect.getsource(self._registry[name])
```

Comment:
Are there other options than this? Feels a bit hacky...

Reply:
I tried quite a few things, but none were as reliable. However, I found an alternative that might be a bit nicer.

The issue arises when adding components to `urlpatterns`. This re-registers the components and, in the case of Components used as views, they're treated as if they are different classes, due to how the packages are imported (or at least that's what I've been able to gather so far).

I tried using:

- `__name__`: But this will replace a component incorrectly if they share the same instance name.
- `__dict__`: For the same component `__module__`, and functions memory addresses aren't the same. So it raises an error about multiple components being registered with the same name. See example: https://pastebin.com/J6SH6eU9
- `__qualname__`: Has the same issue as `__dict_` as qualname (e.g., `greeting.Greeting` vs. `components.greeting.Greeting` are considered different).

I've found that the most reliable way to compare if two components are the same is by comparing the source codes, or using the location of the file + class name as follows:

1. Adding a `class_hash` to `Component`, based on the file the class is and its name:
```python
@classmethod
@property
class_hash(cls):
     return hash(str(inspect.getfile(cls)) + str(cls.__name__))
```

2. Doing the comparison in `component_registry.py` like this:
```python
    def register(self, name=None, component=None):
        existing_component = self._registry.get(name)
        if (
            existing_component
            and existing_component.class_hash != component.class_hash
        ):
            raise AlreadyRegistered(
                'The component "%s" has already been registered' % name
            )
        self._registry[name] = component
```

What do you think? Any thoughts on better/alternative approaches?